src/ocean/text/convert/Formatter.d

```diff
@@ -478,7 +478,21 @@ private void handle (T) (T v, FormatInfo f, scope FormatterSink sf, scope ElemSi
         nullWrapper(&v, se(v.toString(), f), se("null", f));
     else static if (is(T == interface))
         handle!(Object)(cast(Object) v, f, sf, se);
-
+    else static if (isForwardRange!T && !isInfinite!T)
+    {
+        Flags old_flag = f.flags;
+        f.flags |= Flags.Nested;
+        sf("[");
```

Comment:
You need to call `.save`. That's how forward ranges work. You essentially write code like this:

```D
auto range = v.save;
while (!range.empty)
{
    range.popFront();
}
```

https://tour.dlang.org/tour/en/basics/ranges
https://dlang.org/phobos/std_range_primitives.html#isForwardRange

Unfortunately the Korean version is outdated and doesn't describe forward ranges on https://tour.dlang.org/tour/kr/basics/ranges.

Reply:
Well, I finally get the meaning of `**not to consume**` now